Analyze constants to remove unnecessary branches

JamesIry opened the pull request:
This commit adds analysis and optimization of constants to remove
unnecessary branches. It uses abstract interpretation to determine
what constant(s) a particular stack slot or variable might or might not
hold at a given spot and uses that knowledge to eliminate branches that
cannot be taken. Its primary goal is null check removal, but it also
works for other constants.

Several tests are modified to include the new optimization phase.

Two new tests are added. One verifies that branching still works as
expected. The other verifies that branches are removed.

JamesIry committed:
Analyze constants to remove unnecessary branches

This commit adds analysis and optimization of constants to remove
unnecessary branches. It uses abstract interpretation to determine
what constant(s) a particular stack slot or variable might or might not
hold at a given spot and uses that knowledge to eliminate branches that
cannot be taken. Its primary goal is null check removal, but it also
works for other constants.

Several tests are modified to include the new optimization phase.

Two new tests are added. One verifies that branching still works as
expected. The other verifies that branches are removed.

JamesIry commented:
review @retronym

retronym commented:
Neat, this will help me bring SI-7172 home, too. I'll take a closer look at this patch in tomorrow.

magarciaEPFL commented:
@JamesIry , I haven't looked in detail at your implementation , just at the description of its goals (constant folding, including discarding branches not taken after null checks).

It appears that performing Abstract Interpretation sequentially on ICode CFGs will invariably be slower than at the ASM level (GenASM prepares ASM trees anyway).

It's clear the optimization should be done, to get rid of redundant code. However, there are advantages to performing this (very same) optimization at the ASM level, for example via:

  https://github.com/magarciaEPFL/scala/blob/GenRefactored2/src/asm/scala/tools/asm/optimiz/ConstantFolder.java
  https://github.com/magarciaEPFL/scala/blob/GenRefactored2/src/asm/scala/tools/asm/optimiz/NullnessPropagator.java

The main advantage being parallelism.

Is there any phase ordering advantage to having an ICode phase for the above?

JamesIry commented:
@magarciaEPFL There is an advantage right now: namely this optimization only replaces conditional branches with jumps, it doesn't remove any blocks nor analyze what code it renders useless. Later phases do dead code elimination, unreachable block elimination, peephole optimization, jump collapsing, etc.  As you know there's nothing preventing all those things from being done on ASM, but we don't have that stuff just yet.

JamesIry commented:
@retronym This version doesn't yet understand that on the true side of x.isInstanceOf[Something] x can't be null (until mutated). That's in its list of TODOs.

magarciaEPFL commented:
@JamesIry

> As you know there's nothing preventing all those things from being done on ASM, but we don't have that stuff just yet.

The optimizations you mention are available at https://github.com/magarciaEPFL/scala/tree/GenRefactored2/src/asm/scala/tools/asm/optimiz

They are not tied to the new optimizer, just to ASM. The contract of each optimization is documented on each file header.

To use those optimizations from GenASM, basically instead of subclassing `asm.ClassWriter` as is done now, `asm.tree.ClassNode` should be used (both `ClassNode` and `ClassWriter` are behavioral subtypes of `ClassVisitor`). GenASM will not notice, and then (a chosen subset of) `asm.optimiz` optimizations can be applied.

magarciaEPFL commented:
Let's see what `NullnessPropagator` does:

```
    package scala.tools.asm.optimiz;

    /**
     *  Infers null resp. non-null reaching certain program points, using that information to:
     *
     *    (a) simplify control-flow when a conditional jump will always be taken or avoided,
     *        for three classes of instructions:
     *          (a.1) IF_ACMPEQ, IF_ACMPNE
     *          (a.2) IFNULL,    IFNONNULL
     *          (a.3) INSTANCEOF
     *
     *    (b) simplify an ALOAD for a local-var known to contain null (ACONST_NULL replaces it).
     *        This might enable further reductions (e.g., dead-store elimination).
     *
     *    (c) Scala unbox of null is replaced by a 0 load of the appropriate sort (I, L, F, or D).
     *
     *  @author  Miguel Garcia, http://lamp.epfl.ch/~magarcia/ScalaCompilerCornerReloaded/
     *  @version 1.0
     */
    public class NullnessPropagator {
```

An instanceof bytecode instruction whose outcome is known is rewritten as follows:

```
case Opcodes.INSTANCEOF:
    if(frame.getStackTop().isNull()) {
        // drop ref, ICONST_0
        TypeInsnNode iof = (TypeInsnNode)insns[i];
        mnode.instructions.insert(iof, new InsnNode(Opcodes.ICONST_0));
        mnode.instructions.insert(iof, Util.getDrop(1));
        mnode.instructions.remove(iof);
        changed = true;
    }
    break;
```

JamesIry commented:
@magarciaEPFL pull requests welcome

magarciaEPFL commented:
@JamesIry What's that supposed to mean?

JamesIry commented:
@magarciaEPFL Just that: please don't leave that stuff lying around in a private fork. Start shoveling in! We want it! It's much easier for you to generate bite sized, self consistent pull requests from your code than it is for me to troll through a large, unknown code base trying to find and package up nuggets. I may well do that at some point, but not today.

retronym commented on src/compiler/scala/tools/nsc/backend/opt/ConstantOptimization.scala:
```diff
@@ -0,0 +1,639 @@
+          val in1 = in peek 0
+          val newStuff = tags zip labels filter { case (tagSet, _) => canSwitch(in1, tagSet) }
+          val (reachableTags, reachableNormalLabels) = (tags zip labels filter { case (tagSet, _) => canSwitch(in1, tagSet) }).unzip
+          val reachableLabels = if (labels.size > tags.size) {
```
Could use `lengthCompare`.

retronym commented on src/compiler/scala/tools/nsc/backend/opt/ConstantOptimization.scala:
```diff
@@ -0,0 +1,639 @@
+        }
+
+        if (result.size == 1) (result, List.fill(toDrop)(DROP(kind)) :+ JUMP(result.keySet.head))
+        else (result, inst :: Nil)
```
Push the `if` inside `(result, ... )`.

retronym commented on src/compiler/scala/tools/nsc/backend/opt/ConstantOptimization.scala:
```diff
@@ -0,0 +1,639 @@
+        in drop drops
+
+      case CALL_PRIMITIVE(_) =>
+        in drop inst.consumed push UNKNOWN
```
Perhaps use `in drop inst.consumed` all over, rather than `drop <literal>`.

retronym commented on src/compiler/scala/tools/nsc/backend/opt/ConstantOptimization.scala:
```diff
@@ -0,0 +1,639 @@
+        // TODO if the top of stack is guaranteed null then we can eliminate this IS_INSTANCE check and 
+        // replace with a constant false, but how often is a knowable null checked for instanceof?
+        // TODO we could track type information and statically know to eliminate IS_INSTANCE
+        // but that's probably not a huge win
```
Such cases might be interesting under specialization; that's the subject of [SI-7172](https://issues.scala-lang.org/browse/SI-7172). We could also handle this earlier by emitting `if (x eq null) false else [[x.tpe <:< checkTp]]` rather than `x.isInstanceOf[checkTp]` in `Erasure`, as we have all the type information at hand there. But I guess that doesn't profit from refined type information that might be gleaned on subsequent iterations of this optimization process.

JamesIry replied:
That layout would also make it harder to eliminate the null check. With the current x.isInstanceOf[checkTp] it's easy enough to eliminate the null, I just need to do it.

retronym commented on src/compiler/scala/tools/nsc/backend/opt/ConstantOptimization.scala:
```diff
@@ -0,0 +1,639 @@
+       * True if this constant has the same representation (and therefore would compare true under eq) as another constant
+       */
+      override def equals(other: Any) = (other match {
+        case oc @ Const(o) => (this eq oc) || (if (this.isIntAssignable && oc.isIntAssignable) this.toInt == oc.toInt else c.value == o.value)
```
I think you'll need to treat `Double.NaN` here. Reuse `Constant#equalHashValue`.

JamesIry replied:
Can't do that because NaN != NaN and I don't want to eliminate the branch in  'if (NaN != NaN) println("Blame IEEE")'

But UGH. NaN means I have to eliminate the shortcircuit check on pointer equality in that line for the same reason I can't use equalsHashValue.

I think I have to eliminate a couple of other pointer equality checks as well.

retronym replied:
Whichever way you slice it, it would be good to see unit tests for equals/hashCode consistency of this class, in addition to the functional tests like `blame-aye-triple-eee.scala`.

JamesIry committed:
Cleanup of constant optimization

This commit cleans up constant optimization from the review of
https://github.com/scala/scala/pull/2214 .

* drops are done using the instruction's consumed count rather than a
  numeric literal
* drops are moved into one common method in the main instruction
  interpreter
* One instance of x.length > y.length is replaced with
  x.lengthCompare(y.length) > 0
* NaN is dealt with by treating it as an UNKNOWN
* A test is added to make sure NaN semantics aren't broken.
* The constant-optmization test is improved with tests for switch
  statements

paulp commented:
Can someone summarize the state of the pull request?

JamesIry commented:
Waiting on second round of feedback from @retronym

JamesIry commented:
I should rephrase because that sounds like @retronym is being slow. He's not. My updates based on his first round review went in this morning. The PR is now waiting for a second round or a thumbs up.

paulp commented:
So Slowy McSlowerson is slowing everything down? Gotcha. Guess we have no choice but to hunker down and wait him out.

retronym commented:
Two more comments.
- `in drop inst.consumed` should be factored out to a method.
- Can you comment on the apparent duplication in effort between these optimizations (against icode), and Miguel's work (against bytecode.)? My read is that: "icode is still the reality on this branch, we need these optimizations in place ASAP, we're happy if they are subsumed (even shortly) down the track, but in any case the test cases will live on.". Which would Sound Good To Me.

Which is reasonable enough to me. I also think that this exercise will help us to review and test Miguel's optimizations, which is also a win.

BTW, GitHub doesn't notify me that new commits have been pushed to a PR; it's best to add a comment to expediate things :)

JamesIry commented:
This optimization works best if dead code elimination, peephole optimization, unreachable branch elimination, dead variable elimination, and block collapsing all come after it. It will take awhile to incorporate all that from Miguel's branch. We can envision a process where ASM based optimizations are brought in from back to front, one at a time, subsuming icode based ones. Eventually that would mean this one is superfluous - or if this one does something that the new back end doesn't then the logic would have to be ported.

retronym commented:
Thanks for the explanation. LGTM, a pleasure to read. But this area of the compiler isn't really in my wheelhouse so I'd love if @magarciaEPFL could also review this for any correctness problems.

JamesIry commented:
If Miguel doesn't have time maybe it would cheer up @gkossakowski to take a look or @dragos can put on his old optimization hat.

magarciaEPFL commented:
I'm a bit under pressure with time, yes. But I promise I'll keep watching this space :)

gkossakowski commented:
I don't think I'll be able to do code reviews for another week. Sorry.

On Wednesday, March 13, 2013, James Iry wrote:

> If Miguel doesn't have time maybe it would cheer up @gkossakowskihttps://github.com/gkossakowskito take a look or
> @dragos https://github.com/dragos can put on his old optimization hat.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/scala/scala/pull/2214#issuecomment-14844721
> .

## 

Grzegorz Kossakowski
Scalac hacker at Typesafe http://www.typesafe.com/
twitter: @gkossakowski http://twitter.com/gkossakowski
github: @gkossakowski http://github.com/gkossakowski

JamesIry commented:
I've got more work to do against this. Based on @retronym's LGTM and a conversation with @adriaanm I'm merging. If any of the afore mentioned cast of characters wants to do an after-the-fact review I'm willing to incorporate changes then.
